Aby uzykać współczynnik beta dla trzech wybranych przez nas spółek: CCC, KGHM oraz Pekao wpierw musimy wyliczyć zwrot rynkowy. Pomoże nam w tym WIG20, który reprezentuje 20 największych i najbardziej płynnych spółek notowanych na warszawskiej giełdzie. 
We wszystkich notowaniach korzystamy z okresu od 03.01.2005 do 29.12.2023

In [95]:
import pandas as pd
import numpy as np

In [96]:
data_wig20 = pd.read_csv("data/wig20_2005.csv")
data_wig20['Dzienny zwrot rynku'] = (data_wig20['Zamkniecie']-data_wig20['Otwarcie'] / data_wig20['Otwarcie'])
data_wig20['Dzienny zwrot rynku']

0       1965.69
1       1953.80
2       1909.33
3       1896.29
4       1902.83
         ...   
4749    2347.67
4750    2345.70
4751    2370.28
4752    2355.32
4753    2341.99
Name: Dzienny zwrot rynku, Length: 4754, dtype: float64

Mamy wyliczone dzienne zwroty rynkowe, obliczmy teraz współczynniki beta dla każdej z trzech spółek:

In [97]:
data_kghm = pd.read_csv("data/kgh_2005.csv")
data_kghm['Dzienny zwrot'] = (data_kghm['Zamkniecie']-data_kghm['Otwarcie'] / data_kghm['Otwarcie'])
data_kghm.head()

Data  Otwarcie  Najwyzszy  Najnizszy  Zamkniecie       Wolumen  \
0  2005-01-03  10.27210   10.43490   10.20380    10.27210  1.351049e+06   
1  2005-01-04  10.23550   10.27210   10.07270    10.10750  2.908545e+06   
2  2005-01-05   9.71454    9.71454    9.48536     9.58551  6.020221e+06   
3  2005-01-06   9.48536    9.48536    9.15794     9.32069  3.856970e+06   
4  2005-01-07   9.32069    9.58551    9.12607     9.45358  2.093734e+06   

   Dzienny zwrot  
0        9.27210  
1        9.10750  
2        8.58551  
3        8.32069  
4        8.45358

In [98]:
covariance_kghm = np.cov(data_kghm['Dzienny zwrot'], data_wig20['Dzienny zwrot rynku'], ddof=0)[0][1]
covariance_kghm


-3928.1631455415736

In [99]:
market_variance = np.var(data_wig20['Dzienny zwrot rynku'], ddof=0)
market_variance


217755.54110027436

In [100]:
beta_kghm = covariance_kghm / market_variance
beta_kghm

-0.0180393257764802

In [101]:
data_pekao = pd.read_csv("data/peo_2005.csv")
data_pekao['Dzienny zwrot'] = (data_pekao['Zamkniecie']-data_pekao['Otwarcie'] / data_pekao['Otwarcie'])
data_pekao.head()

Data  Otwarcie  Najwyzszy  Najnizszy  Zamkniecie        Wolumen  \
0  2005-01-03   56.8084    58.8612    56.8084     58.2476  376792.311012   
1  2005-01-04   57.4235    58.0435    57.2176     57.8319  159300.776621   
2  2005-01-05   57.2176    57.6333    55.7800     55.7800  429800.893829   
3  2005-01-06   55.5742    56.8084    55.1601     55.1601  377216.003060   
4  2005-01-07   55.1601    56.1893    54.9551     55.1601  273211.994084   

   Dzienny zwrot  
0        57.2476  
1        56.8319  
2        54.7800  
3        54.1601  
4        54.1601

In [102]:
covariance_pekao = np.cov(data_pekao['Dzienny zwrot'], data_wig20['Dzienny zwrot rynku'], ddof=0)[0][1]
covariance_pekao

4883.814100413982

In [103]:
beta_pekao = covariance_pekao / market_variance
beta_pekao

0.022427967048448294

In [104]:
data_ccc = pd.read_csv("data/ccc_2005.csv")
data_ccc.head()
len(data_ccc)

4736

In [105]:
data_ccc.head()

Data  Otwarcie  Najwyzszy  Najnizszy  Zamkniecie  Wolumen
0  2005-01-03    7.6302     7.6725     7.5876      7.6725     8145
1  2005-01-04    7.6725     7.6725     7.6302      7.6302     4845
2  2005-01-05    7.5876     7.5876     7.3493      7.3918     3579
3  2005-01-06    7.3918     7.4346     7.3918      7.3918    12006
4  2005-01-07    7.3918     7.5535     7.3918      7.3918     4235

Dla spółki CCC nie mamy notowań dla wszystkich dni, w których WIG20 było notowane, musimy uzupełnić kalendarz o brakujące dni, wypełniając średnią puste komórki

In [106]:
# Wyświetl unikalne daty dla ramki danych data_wig20
unikalne_daty_wig20 = data_wig20['Data'].unique()
print("Unikalne daty dla WIG20:")
print(unikalne_daty_wig20)

# Wyświetl unikalne daty dla ramki danych data_ccc
unikalne_daty_ccc = data_ccc['Data'].unique()
print("\nUnikalne daty dla CCC:")
print(unikalne_daty_ccc)

Unikalne daty dla WIG20:
['2005-01-03' '2005-01-04' '2005-01-05' ... '2023-12-27' '2023-12-28'
 '2023-12-29']

Unikalne daty dla CCC:
['2005-01-03' '2005-01-04' '2005-01-05' ... '2023-12-27' '2023-12-28'
 '2023-12-29']


In [107]:
# Znajdź brakujące daty dla CCC
brakujace_daty_ccc = set(unikalne_daty_wig20) - set(unikalne_daty_ccc)
print("Brakujące daty dla CCC:", brakujace_daty_ccc)

Brakujące daty dla CCC: {'2008-09-11', '2009-02-16', '2008-09-08', '2010-06-28', '2008-07-07', '2008-09-15', '2008-09-16', '2010-06-25', '2010-02-23', '2008-05-09', '2008-12-30', '2008-09-05', '2011-06-16', '2008-09-10', '2008-05-12', '2008-03-20', '2008-05-23', '2008-08-13'}


In [108]:
len(data_wig20)

4754

In [109]:
len(data_ccc)

4736

Widzimy, że data_cc ma 4736 wierszy, a data_wig20 4754. Wyznaczyłam brakujące daty, trzeba je jakoś wprowadzić do datasetu ccc, uzupełniając kolumny o średnią (lub interpolację liniową).
Gdy już oba datasety będą mieć równą liczbę wierszy, możemy przejść do wyliczenia współczynnika bety.